src/hidpp20.c

```diff
@@ -2073,49 +2064,54 @@ int hidpp20_onboard_profiles_write(struct hidpp20_device *device,
 				   unsigned int index,
 				   struct hidpp20_profiles *profiles_list)
 {
-	union hidpp20_internal_profile pdata = {0};
-	uint8_t *data = pdata.data;
+	union hidpp20_internal_profile *pdata;
+	uint8_t *data;
+	uint16_t sector_size = profiles_list->sector_size;
+	uint16_t sector = index + 1;
 	struct hidpp20_profile *profile = &profiles_list->profiles[index];
 	unsigned i;
 	int rc;
 
 	if (index >= profiles_list->num_profiles)
 		return -EINVAL;
 
+	data = hidpp20_onboard_profiles_allocate_sector(profiles_list);
+	pdata = (union hidpp20_internal_profile *)data;
+
 	rc = hidpp20_onboard_profiles_find_and_read_profile(device,
 							    index,
+							    profiles_list->sector_size,
 							    data,
 							    profiles_list->num_rom_profiles);
 	if (rc < 0)
-		return rc;
+		goto out;
 
 	rc = hidpp20_onboard_profiles_set_enable_profile(device, index, profiles_list);
 	if (rc < 0)
-		return rc;
+		goto out;
 
-	pdata.profile.report_rate = 1000 / profile->report_rate;
-	pdata.profile.default_dpi = profile->default_dpi;
-	pdata.profile.switched_dpi = profile->switched_dpi;
+	pdata->profile.report_rate = 1000 / profile->report_rate;
+	pdata->profile.default_dpi = profile->default_dpi;
+	pdata->profile.switched_dpi = profile->switched_dpi;
 
 	for (i = 0; i < 5; i++) {
-		pdata.profile.dpi[i] = hidpp_cpu_to_le_u16(profile->dpi[i]);
+		pdata->profile.dpi[i] = hidpp_cpu_to_le_u16(profile->dpi[i]);
 	}
 
 	for (i = 0; i < profiles_list->num_leds; i++)
-		hidpp20_onboard_profiles_write_led(&pdata.profile.leds[i], &profile->leds[i]);
+		hidpp20_onboard_profiles_write_led(&pdata->profile.leds[i], &profile->leds[i]);
 
-	hidpp20_buttons_from_cpu(profile, pdata.profile.buttons, profiles_list->num_buttons);
+	hidpp20_buttons_from_cpu(profile, pdata->profile.buttons, profiles_list->num_buttons);
 
-	memcpy(pdata.profile.name.txt, profile->name, sizeof(profile->name));
+	memcpy(pdata->profile.name.txt, profile->name, sizeof(profile->name));
 
-	rc = hidpp20_onboard_profiles_write_page(device, index + 1, data);
+	rc = hidpp20_onboard_profiles_write_sector(device, sector, sector_size, data);
 	if (rc < 0)
-		return rc;
+		goto out;
 
-	if (rc != sizeof(pdata.data))
-		return -EIO;
-
-	return 0;
+ out:
+	free(data);
+	return rc < 0 ? rc : 0;
```

Comment:
these code paths seems like they could be improved using the `_cleanup_free_` helpers, but this isn't something I'll insist on